hasktorch/src/Torch/Optim.hs

```diff
@@ -11,13 +11,26 @@ import Torch.Autograd
 import Torch.NN
 
 type LearningRate = Tensor
+type Loss = Tensor -- TODO - consider newtype wrapping?
 newtype Gradients = Gradients [Tensor] deriving Show
 
 grad' t p = Gradients (grad t p)
 
 class Optimizer o where
     step :: LearningRate -> Gradients -> [Tensor] -> o -> ([Tensor], o)
 
+-- | run a single iteration of an optimizer, returning new parameters and updated optimizer state
+runStep :: (Show p, Parameterized p, Optimizer o) =>
+        p -> o -> Loss -> LearningRate -> IO ([Parameter], o)
+runStep paramState optState lossValue lr = do
```

Comment:
ah, I defined the same function for the typed optimizers :)

Reply:
This was originally `runIter` in the optimizers example and was a bit messier w/ its responsibilities (had the loss function passed in as a HoF instead of the value, did logging which required iteration information be passed in), rejiggered it to have the same shape as your module's function.